
igvtools-index
==============
*bioinformatics*

Documentation
-------------

URL
******
`https://software.broadinstitute.org/software/igv/igvtools_commandline#index <https://software.broadinstitute.org/software/igv/igvtools_commandline#index>`_

Docstring
*********
The igvtools utility provides a set of tools for pre-processing data files. 
    File names must contain an accepted file extension, e.g. test-xyz.bam. Tools include:

    toTDF: 
        Converts a sorted data input file to a binary tiled data (.tdf) file. Used to preprocess large datasets 
        for improved IGV performance. Supported input file formats: .cn, .gct, .igv, .res, .snp, and .wig
        Note: This tool was previously known as _tile_ 
    
    count:
        Computes average alignment or feature density for over a specified window size across the genome and 
        outputs a binary tiled data .tdf file, text .wig file, or both depending on inputs.
        Used to create a track that can be displayed in IGV, for example as a bar chart.
        Supported input file formats: .aligned, .bam, .bed, .psl, .pslx, and .sam
    
    index
        Creates an index file for an ASCII alignment or feature file.
        Index files are required for loading alignment files into IGV, and can significantly improve 
        performance for large feature files. The file must be in sorted order by start position.
        Supported input file formats: .aligned, .bed, .psl, .sam, .bam, and .vcf (v3.2)

    sort
        Sorts the input file by start position. 
        Used to prepare data files for tools that required sorted input files.
        Supported input file formats: .aligned, .bed, .cn, .igv, .psl, .sam, .bam, and .vcf

    Documentation: https://software.broadinstitute.org/software/igv/igvtools
        
    ---------------------------------------------------------
    
    Creates an index for an alignment or feature file. Index files are required for loading alignment files 
    into IGV, and can significantly improve performance for large feature files. Note that the index file 
    is not directly loaded into IGV. Rather, IGV looks for the index file when the alignment or feature file 
    is loaded. This command does not take an output file argument. Instead, the filename is generated by 
    appending ".sai" (for alignments) or ".idx" (for features) to the input filename as IGV relies on this 
    naming convention to find the index . The input file must be sorted by start position (see sort command, below). 

    Supported input file formats are: .sam, .bam, .aligned, .vcf, .psl, and .bed.
    
    Note: The "sai" index is an IGV format, it does not work with samtools or any other application.

*This page was automatically generated*
